### ISSUE TITLE
Loading model with tf.keras.models.load_model not working on multi GPU

### ISSUE THREAD

**Issue by MariaHeuss** [tensorflow/tensorflow#35942]
**System information**
- Custom code
- TensorFlow version 2.1.0
- Python version: 3.7
- GPU model: 4 V100 GPUs on Kubernetes Engine

**Describe the current behavior**
On multi GPU loading the model from a h5 file is not working. 

**Describe the expected behavior**
Saving and reloading the model from a h5 file using model.save and  keras.models.load_model should work on both single and multi GPU.

**Code to reproduce the issue**
``` python 
import tensorflow as tf 
import os
import contextlib
import numpy as np
import tensorflow.keras as keras  

def get_model():
    model = keras.Sequential([
        keras.layers.Flatten(input_shape=(28, 28)),
        keras.layers.Dense(10, activation=tf.nn.softmax)
    ])
    model.compile(optimizer=tf.keras.optimizers.Adam(),
                      loss='sparse_categorical_crossentropy')
    return model

def get_model_path():
    model_dir = '/tmp/m' + str(np.random.randint(0, 1000000))
    os.makedirs(model_dir)
    model_path = os.path.join(model_dir, 'model')
    return model_path + ".h5"

def attempt_save_and_reload(model_path, distributed_training=False):
    fashion_mnist = keras.datasets.fashion_mnist
    (train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
    train_images = train_images / 255.0
    test_images = test_images / 255.0

    with strategy.scope() if distributed_training else contextlib.nullcontext():
        model = get_model()
        model.fit(
            train_images,
            train_labels,
            epochs=1,
        )
        model.save(model_path)
        model = tf.keras.models.load_model(model_path)

if __name__ == '__main__':
    strategy = tf.distribute.MirroredStrategy()
    for distributed_training in [False, True]:
        print('distributed training: ', distributed_training)
        model_path = get_model_path()
        try:
            attempt_save_and_reload(model_path, distributed_training)
        except Exception as e:
            print('Exception raised: \n', e)
        print()
```

**other info/ logs**
I need to use h5 files since saving the optimizer state does not work otherwise (see #33424).  The logs I get are: 
```
INFO:tensorflow:Using MirroredStrategy with devices ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1', '/job:localhost/replica:0/task:0/device:GPU:2', '/job:localhost/replica:0/task:0/device:GPU:3')
distributed training:  False
Train on 60000 samples
60000/60000 [==============================] - 3s 52us/sample - loss: 0.5991

distributed training:  True
Train on 60000 samples
INFO:tensorflow:batch_all_reduce: 2 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:batch_all_reduce: 2 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
60000/60000 [==============================] - 9s 152us/sample - loss: 0.6016
Exception raised: 
 `handle` is not available outside the replica context or a `tf.distribute.Strategy.update()` call.
```

**Comment by lgeiger** [comment]
@ravikyram @ymodak The failure can also be observed in this [colab notebook](https://colab.research.google.com/drive/1nXw4V5LwtxrVNB8XlJ5TaOxbAOXRS1O3).

**Comment by pidajay** [comment]
The loading works if one saves and loads the model outside the distribution context (i.e outside the with strategy.scope() block). Looks like one is not allowed to save or load models when in a cross device context. Is it safe to assume that from tf 2.1+ one can save and load keras models only outside the distribution context?

**Comment by lgeiger** [comment]
@pidajay How would you fit the reloaded model using multiple GPUs then? Could you provide a code example?

**Comment by ckkuang** [comment]
Loading and saving keras models inside a distribution strategy scope as TF SavedModel / Checkpoint format is supported and should work as expected. However loading .h5 format model inside a strategy scope is not yet supported.

However as mentioned in one of the early thread ([#33424](https://github.com/tensorflow/tensorflow/issues/33424)), loading a SavedModel / Checkpoint doesn't restore the optimizer state. Before the issue is fixed, only model weights will be carried over and the training won't be resumed exactly where you left off.

**Comment by CalebEverett** [comment]
This appears not to work across different distribution strategies. I saved a model under a tpu distribution strategy and tried to load it with just cpu and it didn't work, producing the following error below. The model was saved with tf version 2.2.0-dev20200309 (in 2.1 save_model with a tpu distribution strategy didn't work) and tried to load with tf version 2.1.0-rc0.
```
strategy =  tf.distribute.get_strategy()
with strategy.scope():
    model = tf.keras.models.load_model(model_path)
```

Also loading weights appears not to work between the two environments.

```
/opt/conda/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py in load_weights(self, filepath, by_name, skip_mismatch)
    232         raise ValueError('Load weights is not yet supported with TPUStrategy '
    233                          'with steps_per_run greater than 1.')
--> 234     return super(Model, self).load_weights(filepath, by_name, skip_mismatch)
```

**Comment by CalebEverett** [comment]
training on tpu with 2.2.0-dev20200311 and loading under strategy.scope() on a different machine with gpu only runs on the cpu.

tried `strategy = tf.distribute.OneDeviceStrategy('gpu:0')` to no avail.

tried loading with `tf.saved_model.load` and inferencing with `model.signatures['serving_default']` and got the following error.

```
InvalidArgumentError: Cannot assign a device for operation StatefulPartitionedCall/sequential_6/efficientnet-b0/stem_bn/FusedBatchNormV3/AddN: {{node StatefulPartitionedCall/sequential_6/efficientnet-b0/stem_bn/FusedBatchNormV3/AddN}} was explicitly assigned to /job:worker/replica:0/task:0/device:CPU:0 but available devices are [ /job:localhost/replica:0/task:0/device:CPU:0, /job:localhost/replica:0/task:0/device:XLA_CPU:0 ]. Make sure the device specification refers to a valid device.
	 [[StatefulPartitionedCall/sequential_6/efficientnet-b0/stem_bn/FusedBatchNormV3/AddN]] [Op:__inference_signature_wrapper_313855]
```

**Comment by CalebEverett** [comment]
tried loading the model from `tf.keras.models.model_from_json` and got the following error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-72d87b8e325d> in <module>
      2     model_config = json.load(json_file)
      3 
----> 4 model = tf.keras.models.model_from_json(model_config)

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/saving/model_config.py in model_from_json(json_string, custom_objects)
    114   config = json.loads(json_string)
    115   from tensorflow.python.keras.layers import deserialize  # pylint: disable=g-import-not-at-top
--> 116   return deserialize(config, custom_objects=custom_objects)

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/layers/serialization.py in deserialize(config, custom_objects)
    107       module_objects=globs,
    108       custom_objects=custom_objects,
--> 109       printable_module_name='layer')

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py in deserialize_keras_object(identifier, module_objects, custom_objects, printable_module_name)
    371             custom_objects=dict(
    372                 list(_GLOBAL_CUSTOM_OBJECTS.items()) +
--> 373                 list(custom_objects.items())))
    374       with CustomObjectScope(custom_objects):
    375         return cls.from_config(cls_config)

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/sequential.py in from_config(cls, config, custom_objects)
    398     for layer_config in layer_configs:
    399       layer = layer_module.deserialize(layer_config,
--> 400                                        custom_objects=custom_objects)
    401       model.add(layer)
    402     if (not model.inputs and build_input_shape and

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/layers/serialization.py in deserialize(config, custom_objects)
    107       module_objects=globs,
    108       custom_objects=custom_objects,
--> 109       printable_module_name='layer')

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py in deserialize_keras_object(identifier, module_objects, custom_objects, printable_module_name)
    371             custom_objects=dict(
    372                 list(_GLOBAL_CUSTOM_OBJECTS.items()) +
--> 373                 list(custom_objects.items())))
    374       with CustomObjectScope(custom_objects):
    375         return cls.from_config(cls_config)

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/network.py in from_config(cls, config, custom_objects)
    980     """
    981     input_tensors, output_tensors, created_layers = reconstruct_from_config(
--> 982         config, custom_objects)
    983     model = cls(inputs=input_tensors, outputs=output_tensors,
    984                 name=config.get('name'))

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/network.py in reconstruct_from_config(config, custom_objects, created_layers)
   2012   # First, we create all layers and enqueue nodes to be processed
   2013   for layer_data in config['layers']:
-> 2014     process_layer(layer_data)
   2015   # Then we process nodes in order of layer depth.
   2016   # Nodes that cannot yet be processed (if the inbound node

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/network.py in process_layer(layer_data)
   1994       from tensorflow.python.keras.layers import deserialize as deserialize_layer  # pylint: disable=g-import-not-at-top
   1995 
-> 1996       layer = deserialize_layer(layer_data, custom_objects=custom_objects)
   1997       created_layers[layer_name] = layer
   1998 

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/layers/serialization.py in deserialize(config, custom_objects)
    107       module_objects=globs,
    108       custom_objects=custom_objects,
--> 109       printable_module_name='layer')

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py in deserialize_keras_object(identifier, module_objects, custom_objects, printable_module_name)
    360     config = identifier
    361     (cls, cls_config) = class_and_config_for_serialized_keras_object(
--> 362         config, module_objects, custom_objects, printable_module_name)
    363 
    364     if hasattr(cls, 'from_config'):

~/anaconda3/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py in class_and_config_for_serialized_keras_object(config, module_objects, custom_objects, printable_module_name)
    319   cls = get_registered_object(class_name, custom_objects, module_objects)
    320   if cls is None:
--> 321     raise ValueError('Unknown ' + printable_module_name + ': ' + class_name)
    322 
    323   cls_config = config['config']

ValueError: Unknown layer: FixedDropout
```

**Comment by murdav** [comment]
Hello All,

Same issue using Colab (TF 2.2 RC2).
My test-model run and then it was saved without implementing TPU or GPU.  I wanted to test the behaviour. Now loading it, I have:

```
new_model = tf.keras.models.load_model('model.h5')
new_model.summary()
```

```
    319   cls = get_registered_object(class_name, custom_objects, module_objects)
    320   if cls is None:
--> 321     raise ValueError('Unknown ' + printable_module_name + ': ' + class_name)
    322 
    323   cls_config = config['config']

ValueError: Unknown layer: FixedDropout

```

**Comment by ckkuang** [comment]
@CalebEverett For saving and loading across different strategies, can you try it in tf 2.2? It might haven been fixed. BTW when using load_weights / checkpoint.restore, if the model that you are restoring into was created in a scope (potentially different than the save scope) then you need to call restore within that scope.

**Comment by ckkuang** [comment]
@murdav can you post the colab link that leads to this error?

**Comment by murdav** [comment]
@ckkuang sorry for later reply. It was my fault. Basically I'm using https://github.com/qubvel/efficientnet and I have to import the `load_model` like:

```
import efficientnet.tfkeras
from tensorflow.keras.models import load_model

model = load_model('path/to/model.h5')
```

Otherwise and logically I have `ValueError: Unknown layer: FixedDropout`.

Davide

**Comment by laura-rieger** [comment]
Having the same issue as @CalebEverett 
It seems to be that loading expects the input shape to be a list and it is a tupel?

**Comment by ymodak** [comment]
@MariaHeuss This is fixed with TensorFlow 2.2 Thanks!

**Comment by efournie** [comment]
From: Chenkai KuangSent: jeudi 16 avril 2020 23:34To: tensorflow/tensorflowCc: Eric Fournié; ManualSubject: Re: [tensorflow/tensorflow] Loading model with tf.keras.models.load_model not working on multi GPU (#35942) @CalebEverett For saving and loading across different strategies, can you try it in tf 2.2? It might haven been fixed. BTW when using load_weights / checkpoint.restore, if the model that you are restoring into was created in a scope (potentially different than the save scope) then you need to call restore within that scope.—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or unsubscribe.
